### PR TITLE
Throw error if creating too many connections on one thread

### LINKED ISSUE #1102: Too many connections on one thread leads to erroneous results
Creating more than 2²⁷ = 134217728 connections with one synapse type on one thread leads to erroneous results when simulating. This is because the `Target` class uses 27 bits for storing the local connection index, `lcid`. New connections beyond those 134217728 therefore get the wrong `lcid`, leading to wrong results.

There are asserts in the `Target` class that check if `lcid` gets too high, but they never trigger, and no other messages that something might go wrong are given when trying to create more connections than can be handled.

See also #1043.

### PR DESCRIPTION
**Pull request by hakonsbm**
There is currently a limit of 2²⁷ connections per synapse per thread, due to having 27 bits for the `lcid` in the `Target` class. With this PR an exception is thrown if attempting to create more connections than `Target` can handle.

Fixes #1102.

### CODE REVIEW

File: nestkernel/connection_manager.cpp [hunk 1]
```diff
@@ -685,6 +685,10 @@ nest::ConnectionManager::increase_connection_count( const thread tid,
     num_connections_[ tid ].resize( syn_id + 1 );
   }
   ++num_connections_[ tid ][ syn_id ];
+  if ( num_connections_[ tid ][ syn_id ] >= 1 << 27 )
```
Comment (clinssen) [inline nestkernel/connection_manager.cpp]: Can the expression (1 << 27) be changed to `MAX_LCID` after merging #1088? See: https://github.com/nest/nest-simulator/pull/1088/files#diff-42615d7378f1abcdbe8ff2af37737101R78 Thanks!

Comment (hakonsbm) [reply nestkernel/connection_manager.cpp]: Yes, I'll update it once #1088 is merged.

Comment (heplesser) [reply nestkernel/connection_manager.cpp]: Is this comparison strictly safe, i.e., will everything work with 2^27 connections, or is the limit rather 2^27-1?

File: nestkernel/connection_manager.cpp [hunk 2]
```diff
@@ -685,6 +685,10 @@ nest::ConnectionManager::increase_connection_count( const thread tid,
...
+  {
+    throw KernelException( "Too many connections per thread for synapse." );
```
Comment (heplesser) [inline nestkernel/connection_manager.cpp]: I think this message is not very informative for the end user. How about something like `"Too many connections: at most ??? connections supported per virtual process and synapse model."`? The number can be inserted using `compose()` from `libnestutil`.

File: nestkernel/connection_manager.cpp (hunk 1 above)
Comment (hakonsbm) [reply nestkernel/connection_manager.cpp]: @heplesser You are right, I have changed it to 2^27-1.

File: nestkernel/connection_manager.cpp (hunk 2 above)
Comment (hakonsbm) [reply nestkernel/connection_manager.cpp]: I have made the error message more informative.

**Comment by heplesser** [comment]
@hakonsbm One important thing: Once #1088 is in place, the hard-coded `27` in the code introduced here needs to be replaced with the proper compile-time variable. Can you create a PR against jakobj:feature/cmake-target-customization to ensure this?

**Comment by hakonsbm** [comment]
@heplesser I'll take care of it when this is merged.

**Comment by terhorstd** [comment]
marked this as blocked till #1088 is in.

**Pull request merged**